### PR TITLE
bpo-36564: Fix infinite loop in email folding logic

### PR DESCRIPTION
**Pull request by pganssle**
As far as I can tell, this infinite loop would be triggered if:

1. The value being folded contains a single word (no spaces) longer than
   max_line_length
2. The max_line_length is shorter than the encoding's name + 9
   characters.

[bpo-36564](https://bugs.python.org/issue36564): https://bugs.python.org/issue36564

https://bugs.python.org/issue36564

### CODE REVIEW

**Comment by pganssle** [comment]
I believe this will cause merge conflicts with #12020. I have subscribed to that PR and will rebase if it is merged first.

File: Lib/email/_header_value_parser.py
Comment (pganssle) [inline Lib/email/_header_value_parser.py]: I picked `ValueError` here, but if there's a better error type let me know, I'm not super familiar with the `email` module.

**Comment by pganssle** [comment]
@maxking @msapiro @warsaw I have rebased this after #12020 was merged, and it is ready for review. I figured I'd tag Abhilash and Mark on this, as potential new `email` maintainers.

File: Lib/email/_header_value_parser.py
Comment (maxking) [reply Lib/email/_header_value_parser.py]: We should probably use `email.errors.HeaderParseError` instead of `ValueError`. 

It would be nice if you could use single quotes for consistency in the module. 

Also, the error message seems a bit vague given that it is not too short to fit *any* value, but too short to hold encoded word. An ascii value would be just fine with the line length).  And, `Maximum length` is a hard to make sense when coming from an exception message. `max_line_length` is more greppable to me.
What do you think about this error?

 `max_line_length is too small to fit an encoded word`.

File: Lib/test/test_email/test_policy.py
Comment (maxking) [inline Lib/test/test_email/test_policy.py]: Slight change to the comment to make it a bit more clear why this bug is trigerred when there is a single world longer than `maxlen`.

```
# Note: This is only triggered when there is a single word longer than
# max_line_length. This is because when we encounter a word longer than max_line_length,
# it is broken down into encoded words to fit max_line_length. If the max_line_length isn't
# large enough to even  contain RFC 2047 chrome (`?=<charset>?q??=`), we fail.
# Hence the 1234567890 at the end of this whimsical subject.
```

Just to make it clear that Header can have words longer than max line length, as  long as `max_line_length > 7 + len(charset) + 1`.

**Review by maxking — COMMENTED** [review]
Overall, the changes look good to me. I have added some inline comments for error message and slight change to a comment.

Thanks @pganssle !

File: Lib/test/test_email/test_policy.py
Comment (warsaw) [reply Lib/test/test_email/test_policy.py]: (There's an extra space in your comment: "even  contain"

Comment (warsaw) [reply Lib/test/test_email/test_policy.py]: `HeaderParseError` seems right to me too, but maybe @bitdancer has a different opinion?

File: Lib/email/_header_value_parser.py
Comment (warsaw) [reply Lib/email/_header_value_parser.py]: +1 for `HeaderParseError` but @bitdancer has more up-to-date perspective than I might have.

File: Lib/test/test_email/test_policy.py
Comment (warsaw) [inline Lib/test/test_email/test_policy.py]: Missing period.

**Review by warsaw — CHANGES_REQUESTED** [review]
See inlined comments.

File: Lib/test/test_email/test_policy.py
Comment (msapiro) [reply Lib/test/test_email/test_policy.py]: Actually, the phrase "even  contain RFC 2047 chrome (`?=<charset>?q??=`)" should probably be "even contain RFC 2047 chrome plus one character(`?=<charset>?q?x?=`)"

**Comment by msapiro** [comment]
I agree with @maxking that this is a good change - much better than leaving the looping possibility open with just a comment in the code. I also agree with all of @maxking  review comments with one minor note inline and the thanks to @pganssle for developing this.

File: Lib/email/_header_value_parser.py
Comment (bitdancer) [reply Lib/email/_header_value_parser.py]: No, a ValueError would be more appropriate here, I think, because it is an error in the *value* of maxlen.  This isn't a *parsing* error, it's a *rendering* error.

Now, that said you might want to consider the fact that in `_fold_mime_parameters` I deal with this issue by bumping maxlen to 78 rather than raising an error.  I'm not sure that was the right choice,  but whatever we do, it should probably be made consistent between the two cases.  It is so sad that I never came back to fix that XXX comment I left myself :(

Note: I unfortunately am not too likely to see github comments, I get too much noise from that source...I just caught this one by chance.  If you need my attention please post on the issue.

**Comment by maxking** [comment]
Also @pganssle you may want to change the description, since the first point (The value being folded contains a single word (no spaces) longer than max_line_length) isn't true.

File: Lib/email/_header_value_parser.py
Comment (pganssle) [reply Lib/email/_header_value_parser.py]: > Also, the error message seems a bit vague given that it is not too short to fit _any_ value, but too short to hold encoded word. An ascii value would be just fine with the line length). And, `Maximum length` is a hard to make sense when coming from an exception message. `max_line_length` is more greppable to me.

I agree that "Maximum length" is not going to be a good error, switching to `max_line_length` is a good idea.

Regarding the "encoded word" situation, I guess when I wrote this I had assumed that the RFC chrome was added to every line regardless, but only realized that that wasn't the case through testing. I'm still not entirely clear *why* the encoding chrome is unnecessary for shorter words, but I think the suggested comments at least explain *when* the chrome becomes necessary.

**Comment by pganssle** [comment]
Thanks for the reviews everyone. Just as an update, I am planning on making the requested changes soon, but I've just started a new job and I'm still working out some of the IP issues so I've been holding off on any non-trivial OSS work until that's sorted out. Shouldn't be too much longer.

**Comment by pganssle** [comment]
I have made the requested changes; please review again

**Comment by maxking** [comment]
LGTM!

@pganssle It is usually better to not force push and just keep adding more commits as your make changes your existing PR. It makes it a bit easy to remember what changes were made  after the previous review, especially if it has been a while since last review :)

File: Lib/email/_header_value_parser.py
Comment (warsaw) [inline Lib/email/_header_value_parser.py:2858]: Single quotes for module consistency please.

**Review by warsaw — CHANGES_REQUESTED** [review]

File: Lib/email/_header_value_parser.py
Comment (warsaw) [reply Lib/email/_header_value_parser.py:2858]: NM.  I chatted with @maxking and we'll just ignore this.

**Review by warsaw — APPROVED** [review]

**Pull request merged**
